// @dnd-kit-svelte/core/dist/utilities/algorithms/helpers.js
/**
 * Sort collisions from smallest to greatest value
 */
export function sortCollisionsAsc({ data: { value: a } }, { data: { value: b } }) {
    return a - b;
}
/**
 * Sort collisions from greatest to smallest value
 */
export function sortCollisionsDesc({ data: { value: a } }, { data: { value: b } }) {
    return b - a;
}
/**
 * Returns the coordinates of the corners of a given rectangle:
 * [TopLeft {x, y}, TopRight {x, y}, BottomLeft {x, y}, BottomRight {x, y}]
 */
export function cornersOfRectangle({ left, top, height, width }) {
    return [
        {
            x: left,
            y: top,
        },
        {
            x: left + width,
            y: top,
        },
        {
            x: left,
            y: top + height,
        },
        {
            x: left + width,
            y: top + height,
        },
    ];
}
export function getFirstCollision(collisions, property) {
    if (!collisions || collisions.length === 0) {
        return null;
    }
    const [firstCollision] = collisions;
    return property ? firstCollision[property] : firstCollision;
}
