// @dnd-kit-svelte/sortable/dist/hooks/utilities/use-derived-transform.svelte.js
import { getClientRect } from '@dnd-kit-svelte/core';
/*
 * When the index of an item changes while sorting,
 * we need to temporarily disable the transforms
 *
 * TODO: revisit this this situation
 * I don't think this works in svelte - haniel
 */
export function useDerivedTransform(argsFn) {
    const { rect, disabled, index, node } = $derived(argsFn());
    let derivedTransform = $state(null);
    let previousIndex = index;
    $effect(() => {
        if (!disabled && index !== previousIndex && node) {
            const initial = rect;
            if (initial) {
                const current = getClientRect(node, {
                    ignoreTransform: true,
                });
                const delta = {
                    x: initial.left - current.left,
                    y: initial.top - current.top,
                    scaleX: initial.width / current.width,
                    scaleY: initial.height / current.height,
                };
                if (delta.x || delta.y) {
                    derivedTransform = delta;
                }
            }
        }
        if (index !== previousIndex) {
            previousIndex = index;
        }
    });
    $effect(() => {
        if (derivedTransform) {
            derivedTransform = null;
        }
    });
    return {
        get current() {
            return derivedTransform;
        },
    };
}
