// @dnd-kit-svelte/sortable/dist/hooks/defaults.js
import { CSS } from '@dnd-kit-svelte/utilities';
import { arrayMove } from '../utilities/index.js';
export const defaultNewIndexGetter = ({ id, items, activeIndex, overIndex }) => arrayMove(items, activeIndex, overIndex).indexOf(id);
export const defaultAnimateLayoutChanges = ({ containerId, isSorting, wasDragging, index, items, newIndex, previousItems, previousContainerId, transition, }) => {
    if (!transition || !wasDragging) {
        return false;
    }
    if (previousItems !== items && index === newIndex) {
        return false;
    }
    if (isSorting) {
        return true;
    }
    return newIndex !== index && containerId === previousContainerId;
};
export const defaultTransition = {
    duration: 200,
    easing: 'ease',
};
export const transitionProperty = 'transform';
export const disabledTransition = CSS.Transition.toString({
    property: transitionProperty,
    duration: 0,
    easing: 'linear',
});
export const defaultAttributes = {
    roleDescription: 'sortable',
};
